### PR TITLE
Add API's to check versioning status, enable/suspend versioning on a bucket

### PR DESCRIPTION
**Pull request by BigUstad**

### CODE REVIEW

File: Minio.Examples/Cases/DisableVersioning.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
+/*
+ * MinIO .NET Library for Amazon S3 Compatible Cloud Storage, (C) 2017 MinIO, Inc.
```
Comment (kannappanr) [inline Minio.Examples/Cases/DisableVersioning.cs:2]:
```suggestion
 * MinIO .NET Library for Amazon S3 Compatible Cloud Storage, (C) 2020 MinIO, Inc.
```

Comment (kannappanr) [reply Minio.Examples/Cases/DisableVersioning.cs:2]: please make sure that this change is done on all the files that were newly added

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 2]
```diff
@@ -441,5 +441,109 @@ public IObservable<MinioNotificationRaw> ListenBucketNotificationsAsync(string b
...
+                }
+                Uri requestUrl = RequestUtil.MakeTargetURL(this.BaseUrl, this.Secure);
+                SetTargetURL(requestUrl);
+
+                var request = new RestRequest("/" + bucketName, Method.PUT)
+                {
+                    XmlSerializer = new RestSharp.Serializers.DotNetXmlSerializer(),
+                    RequestFormat = DataFormat.Xml
+                };
+                // ``us-east-1`` is not a valid location constraint according to amazon, so we skip it.
```
Comment (poornas) [inline Minio/ApiEndpoints/BucketOperations.cs:497]: this comment can be removed.

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 3]
```diff
@@ -441,5 +441,109 @@ public IObservable<MinioNotificationRaw> ListenBucketNotificationsAsync(string b
...
```
Comment (poornas) [inline Minio/ApiEndpoints/BucketOperations.cs:528]: this comment can be removed.

File: Minio/DataModel/VersioningConfiguration.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
+/*
+ * MinIO .NET Library for Amazon S3 Compatible Cloud Storage, (C) 2017 MinIO, Inc.
```
Comment (poornas) [inline Minio/DataModel/VersioningConfiguration.cs:2]:
```suggestion
 * MinIO .NET Library for Amazon S3 Compatible Cloud Storage, (C) 2020 MinIO, Inc.
```

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 5]
```diff
@@ -441,5 +441,109 @@ public IObservable<MinioNotificationRaw> ListenBucketNotificationsAsync(string b
...
+            try
+            {
+                if (bucketName == null)
+                {
+                    throw new InvalidBucketNameException(bucketName, "bucketName cannot be null");
+                }
+                var request = await this.CreateRequest(Method.GET, bucketName).ConfigureAwait(false);
+                request.AddQueryParameter("versioning","");
+                var response = await this.ExecuteTaskAsync(this.NoErrorHandlers, request, cancellationToken).ConfigureAwait(false);
+            }
```
Comment (poornas) [inline Minio/ApiEndpoints/BucketOperations.cs:462]: you will need to read the versioningconfiguration xml and return the status here.

Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs:485]: Yes, I'd pushed the wrong copy. Thanks..

File: Minio.Examples/Cases/DisableVersioning.cs (hunk 1 above)
Comment (BigUstad) [reply Minio.Examples/Cases/DisableVersioning.cs]: Change done.

File: Minio/DataModel/VersioningConfiguration.cs (hunk 4 above)
Comment (BigUstad) [reply Minio/DataModel/VersioningConfiguration.cs]: I've included this change in all files.

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 6]
```diff
@@ -441,5 +448,130 @@ public IObservable<MinioNotificationRaw> ListenBucketNotificationsAsync(string b
...
+                        Namespace = "http://s3.amazonaws.com/doc/2006-03-01/",
+                        ContentType = config.ToString()
+                    },
+                    RequestFormat = DataFormat.Xml
+                };
+                string body = Regex.Replace(request.XmlSerializer.Serialize(config), @"\t|\n|\r", "");
+                string replace1 = ">\\s+";
+                string replace2 = "\\s+<";
+                body = Regex.Replace(body, replace1, ">");
+                body = Regex.Replace(body, replace2, "<");
```
Comment (poornas) [inline Minio/ApiEndpoints/BucketOperations.cs:526]: @BigUstad , you don't have to do a regex replacement. Just need to marshal the xml and UTF-8 encode. See how SelectObjectContentAsync request is being marshalled as an example.

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 7]
```diff
@@ -441,5 +448,130 @@ public IObservable<MinioNotificationRaw> ListenBucketNotificationsAsync(string b
...
```
Comment (poornas) [inline Minio/ApiEndpoints/BucketOperations.cs:566]: same comment as above - xml needs to be marshalled and UTF-8 encoded.

**Comment by poornas** [comment]
please fix the commit and title of PR to something like "Add API's to check versioning status"

File: Minio/ApiEndpoints/BucketOperations.cs (hunk 6 above)
Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs]: @poornas I did make the change. Thanks for pointing out.

File: Minio/ApiEndpoints/BucketOperations.cs (hunk 7 above)
Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs]: Made the change here too.

File: Minio.Examples/Cases/EnableVersioning.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        // Enable Versioning on a bucket
+        public async static Task Run(MinioClient minio,
+                                     string bucketName = "my-bucket-name")
+        {
+            try
+            {
+                Console.WriteLine("Running example for API: EnableVersioning, ");
+                await minio.EnableVersioningAsync(bucketName);
+                bool enabled = await minio.IsVersioningEnabledAsync(bucketName);
+                Console.WriteLine(( enabled? "Versioning Enabled" : "Unable to Enable Versioning") + " for bucket " + bucketName);
```
Comment (poornas) [inline Minio.Examples/Cases/EnableVersioning.cs:33]:
```suggestion
                Console.WriteLine(( enabled? "Versioning Enabled" : "Versioning could not be enabled") + " for bucket " + bucketName);
```

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 9]
```diff
@@ -88,11 +91,15 @@ public async Task MakeBucketAsync(string bucketName, string location = "us-east-
...
             };
+
             // ``us-east-1`` is not a valid location constraint according to amazon, so we skip it.
             if (location != "us-east-1")
             {
                 CreateBucketConfiguration config = new CreateBucketConfiguration(location);
-                request.AddBody(config);
+                request.XmlSerializer.Namespace = "http://s3.amazonaws.com/doc/2006-03-01/";
+                request.XmlSerializer.ContentType = config.ToString();
+                string body = Regex.Replace(request.XmlSerializer.Serialize(config), @"\t|\n|\r", "");
```
Comment (poornas) [inline Minio/ApiEndpoints/BucketOperations.cs:101]: there is no need to use regex - just serialize XML correctly here and UTF8 encode content

Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs:101]: The changes to MakeBucketsAsync skipped my mind. My bad. Will change right away.

File: Minio.Examples/Cases/EnableVersioning.cs (hunk 8 above)
Comment (BigUstad) [reply Minio.Examples/Cases/EnableVersioning.cs:33]: Fixed output statement.

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 10]
```diff
@@ -30,6 +30,9 @@
 using System.Xml.Linq;
 using System.Xml.Serialization;
 using System.Web;
+using System.Text.RegularExpressions;
```
Comment (kannappanr) [inline Minio/ApiEndpoints/BucketOperations.cs:33]: do you need this?

Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs]: Removed that & other using statements that weren't used.

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 11]
```diff
@@ -441,5 +443,120 @@ public IObservable<MinioNotificationRaw> ListenBucketNotificationsAsync(string b
...
+                }
+                Uri requestUrl = RequestUtil.MakeTargetURL(this.BaseUrl, this.Secure, bucketName);
+                SetTargetURL(requestUrl);
+                VersioningConfiguration config = new VersioningConfiguration(true);
+                var request = new RestRequest("/" + bucketName, Method.PUT)
+                {
+                    XmlSerializer = new RestSharp.Serializers.DotNetXmlSerializer()
+                    {
+                        Namespace = "http://s3.amazonaws.com/doc/2006-03-01/",
+                        ContentType = config.ToString()
```
Comment (poornas) [inline Minio/ApiEndpoints/BucketOperations.cs:511]: ContentType should be "application/xml"

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 12]
```diff
@@ -441,5 +443,120 @@ public IObservable<MinioNotificationRaw> ListenBucketNotificationsAsync(string b
...
+                }
+                Uri requestUrl = RequestUtil.MakeTargetURL(this.BaseUrl, this.Secure);
+                SetTargetURL(requestUrl);
+                VersioningConfiguration config = new VersioningConfiguration(false);
+                var request = new RestRequest("/" + bucketName, Method.PUT)
+                {
+                    XmlSerializer = new RestSharp.Serializers.DotNetXmlSerializer()
+                    {
+                        Namespace = "http://s3.amazonaws.com/doc/2006-03-01/",
+                        ContentType = config.ToString()
```
Comment (poornas) [inline Minio/ApiEndpoints/BucketOperations.cs:547]: ContentType needs to be application/xml

File: Minio/ApiEndpoints/BucketOperations.cs (hunk 11 above)
Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs]: I made the change.

File: Minio/ApiEndpoints/BucketOperations.cs (hunk 12 above)
Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs]: I made the change.

File: Minio.Examples/Cases/EnableVersioning.cs [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+        // Enable Versioning on a bucket
+        public async static Task Run(MinioClient minio,
+                                     string bucketName = "my-bucket-name")
+        {
+            try
+            {
+                Console.WriteLine("Running example for API: EnableVersioning, ");
+                await minio.EnableVersioningAsync(bucketName);
+                VersioningConfiguration vc = await minio.GetVersioningInfoAsync(bucketName);
+                Console.WriteLine();
```
Comment (kannappanr) [inline Minio.Examples/Cases/EnableVersioning.cs:34]: versionConfiguration should be printed here

Comment (BigUstad) [reply Minio.Examples/Cases/EnableVersioning.cs:34]: Removed by mistake. Added back again.

File: Minio/Helper/utils.cs [hunk 14]
```diff
@@ -831,5 +834,56 @@ internal static string getMD5SumStr(byte[] key)
                 {".zip", "application/zip"}
             };
         }
+
+        /// <summary>
+        /// Marshal XML to return marshelelled serialization of the object passed.
```
Comment (poornas) [inline Minio/Helper/utils.cs:839]:
```suggestion
        /// MarshalXML returns marshaled serialization of the object passed. 
```

**Comment by BigUstad** [comment]
Versioning added along with Builder code. Code is merged as of the date of this comment.

**Pull request closed without merging**